demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+        {
+            throw std::runtime_error("PdhSetCounterScaleFactor() failed");
+        }
+
+        coreTimeCounters.push_back(counter);
+    }
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status)
+    {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+}
+
+void CpuMonitor::closeQuery() {
+    PDH_STATUS status = PdhCloseQuery(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCloseQuery() failed");
+    }
+}
+
+void CpuMonitor::enableHistory(std::size_t historySize) {
+    if (0 == historySize) {
+        disableHistory();
+    }
+    else {
+        this->historySize = historySize;
+        historyEnabled = true;
+        if (!lastEnabled) {
+            openQuery();
+        }
+    }
+}
+
+bool CpuMonitor::isHistoryEnabled() const {
+    return historyEnabled;
+}
+
+void CpuMonitor::disableHistory() {
+    historyEnabled = false;
+    if (!lastEnabled) {
+        closeQuery();
+    }
+}
+
+void CpuMonitor::enableLast() {
+    lastEnabled = true;
+    if (!historyEnabled) {
+        historySize = 1;
+        openQuery();
+    }
+}
+
+bool CpuMonitor::isLastEnabled() const {
+    return lastEnabled;
+}
+
+void CpuMonitor::disableLast() {
+    lastEnabled = false;
+    if (!historyEnabled) {
+        closeQuery();
+    }
+}
+
+void CpuMonitor::collectData() {
+    // store CPU Utulization in range [0, 1]
+    PDH_STATUS status;
+    PDH_FMT_COUNTERVALUE DisplayValue;
+
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+
+    std::vector<std::pair<double, double>> idleNonIdleCpuStat;
+    for (std::size_t i = 0; i < coreTimeCounters.size(); i++) {
+        status = PdhGetFormattedCounterValue(coreTimeCounters[i], PDH_FMT_DOUBLE, NULL, &DisplayValue);
+        if (ERROR_SUCCESS != status) {
+            throw std::runtime_error("PdhGetFormattedCounterValue() failed");
+        }
+
+        idleNonIdleCpuStat.emplace_back(DisplayValue.doubleValue, 1 - DisplayValue.doubleValue);
+    }
+
+    std::vector<double> cpuLoad(idleNonIdleCpuStat.size());
+    for (std::size_t i = 0; i < idleNonIdleCpuStat.size(); ++i) {
+        cpuLoad[i] = static_cast<double>(idleNonIdleCpuStat[i].first);
+    }
+
+    if (meanCpuLoad.empty()) {
+        meanCpuLoad = cpuLoad;
+        samplesNumber = 1;
+    } else {
+        for (std::size_t i = 0; i < meanCpuLoad.size(); ++i) {
+            meanCpuLoad[i] = (meanCpuLoad[i] * samplesNumber + cpuLoad[i]) / (samplesNumber + 1);
+        }
+        ++samplesNumber;
+    }
+    cpuLoadHistory.push_back(std::move(cpuLoad));
+    if (cpuLoadHistory.size() > historySize) {
+        cpuLoadHistory.pop_front();
+    }
+}
+
+std::deque<std::vector<double>> CpuMonitor::getLastHistory() const {
+    return cpuLoadHistory;
+}
+
+std::vector<double> CpuMonitor::getMeanCpuLoad() const {
+    return meanCpuLoad;
+}
+#else
```

Comment:
The `#else` suggests that the following code will work on any non-Windows systems, but actually it's quite Linux-specific. You need to change this to `#elif __linux__` and add an actual `#else` with a stub implementation.

Reply:
added